### PR TITLE
Add unique constraint for UM_USER table

### PR DESCRIPTION
**Pull request by sandushi**

### CODE REVIEW

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql [hunk 1]
```diff
@@ -1,3 +1,31 @@
...
+
+CREATE ALIAS ALTER_UM_USER AS $$ void alterUmUser(final Connection conn) throws SQLException {
+    String const_name = "";
+    PreparedStatement ps = conn.prepareStatement("SELECT tc.CONSTRAINT_NAME FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu JOIN INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc ON tc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME WHERE tc.TABLE_NAME ='UM_USER' AND tc.CONSTRAINT_TYPE = 'UNIQUE' AND kcu.COLUMN_NAME ='UM_USER_ID'");
+    ResultSet results =  ps.executeQuery();
+
+    while (results.next()) {
+        const_name  = results.getString("CONSTRAINT_NAME");
+    }
+    if(!const_name.equals("")) {
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql:11]:
```suggestion
    if (!const_name.equals("")) {
```

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql [hunk 2]
```diff
@@ -1,3 +1,31 @@
...
+        ps = conn.prepareStatement("ALTER TABLE UM_USER DROP CONSTRAINT " + const_name);
+        ps.execute();
+    }
+
+    ps = conn.prepareStatement("ALTER TABLE UM_USER ADD UNIQUE(UM_USER_ID)");
+    ps.execute();
+
+    ps = conn.prepareStatement("SELECT tc.CONSTRAINT_NAME FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu JOIN INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc ON tc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME WHERE tc.TABLE_NAME ='UM_USER' AND tc.CONSTRAINT_TYPE = 'UNIQUE' AND kcu.COLUMN_NAME ='UM_USER_NAME'");
+    results =  ps.executeQuery();
+    if(results.next() == false) {
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql:21]:
```suggestion
    if (results.next() == false) {
```

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mssql.sql [hunk 3]
```diff
@@ -1,3 +1,14 @@
-CREATE INDEX INDEX_UM_USERNAME_UM_TENANT_ID ON UM_USER(UM_USER_NAME, UM_TENANT_ID);
+IF EXISTS (SELECT * FROM SYS.OBJECTS WHERE OBJECT_ID = OBJECT_ID(N'[DBO].[UM_USER]') AND TYPE IN (N'U'))
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mssql.sql:1]: What is the purpose of this check?

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mssql.sql [hunk 4]
```diff
@@ -1,3 +1,14 @@
...
+	DECLARE @ConstraintName0 nvarchar(200)
+	SELECT @ConstraintName0 = CONSTRAINT_NAME  FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE WHERE TABLE_NAME='UM_USER' AND COLUMN_NAME='UM_USER_ID'
+	IF @ConstraintName0 IS NOT NULL
+		EXEC('ALTER TABLE UM_USER DROP CONSTRAINT ' + @ConstraintName0);
+	ALTER TABLE UM_USER ADD UNIQUE(UM_USER_ID);
+	DECLARE @ConstraintName1 nvarchar(200)
+	SELECT @ConstraintName1 = CONSTRAINT_NAME  FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE WHERE TABLE_NAME='UM_USER' AND COLUMN_NAME='UM_USER_NAME'
+	IF @ConstraintName1 IS NULL
+		ALTER TABLE UM_USER ADD UNIQUE(UM_USER_NAME, UM_TENANT_ID);
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mssql.sql:10]: Better have this within a BEGIN-END clause to clearly specify the scope of the if block

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mssql.sql (hunk 3 above)
Comment (sandushi) [reply components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mssql.sql:1]: To check the existence of table.

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mysql.sql [hunk 5]
```diff
@@ -1,3 +1,25 @@
-CREATE INDEX INDEX_UM_USERNAME_UM_TENANT_ID ON UM_USER(UM_USER_NAME, UM_TENANT_ID);
+DROP PROCEDURE IF EXISTS ALTER_UM_USER;
+
+DELIMITER $$
+CREATE PROCEDURE ALTER_UM_USER()
+BEGIN
+	IF EXISTS(SELECT * FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA=DATABASE() AND TABLE_NAME='UM_USER') THEN
+	    /* Drop the unique key on (UM_USER_ID,UM_TENANT_ID) with Constraint name UM_USER_ID */
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mysql.sql:7]: Fix indentation

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mysql.sql [hunk 6]
```diff
@@ -1,3 +1,25 @@
...
+		END IF;
+		IF NOT EXISTS (SELECT * FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE WHERE TABLE_SCHEMA=DATABASE() AND TABLE_NAME='UM_USER' AND CONSTRAINT_NAME='UM_USER_NAME') THEN
+			ALTER TABLE `UM_USER` ADD UNIQUE(UM_USER_NAME,UM_TENANT_ID);
+		END IF;
+	END IF;
+END $$
+DELIMITER ;
+
+CALL ALTER_UM_USER();
+
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mysql.sql:22]: Drop the procedure after using. Need to fix it on the cluster script as well.

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql [hunk 7]
```diff
@@ -1,3 +1,31 @@
...
+        ps = conn.prepareStatement("ALTER TABLE UM_USER ADD UNIQUE(UM_USER_NAME,UM_TENANT_ID)");
+        ps.execute();
+    }
+} $$;
+
+CALL ALTER_UM_USER();
+
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql:28]: Drop alias after use.

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/oracle.sql [hunk 8]
```diff
@@ -1,3 +1,56 @@
...
+	databasename VARCHAR2(100);
+
+BEGIN
+	SELECT sys_context('userenv', 'current_schema')
+  	INTO databasename
+  	FROM dual;
+
+  	BEGIN
+		SELECT a.constraint_name
+    	INTO con_name0
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/oracle.sql:14]: Fix indentation

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/oracle.sql [hunk 9]
```diff
@@ -1,3 +1,56 @@
...
+		FROM all_cons_columns a
+			JOIN all_constraints c ON a.owner = c.owner AND a.constraint_name = c.constraint_name
+		WHERE c.constraint_type = 'U' AND a.table_name = 'UM_USER' AND UPPER(a.OWNER) = UPPER(databasename) AND a.column_name='UM_USER_ID';
+
+  		IF con_name0 IS NOT NULL THEN
+      		command := 'ALTER TABLE UM_USER DROP CONSTRAINT ' || con_name0;
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/oracle.sql:20]: Fix indentation

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/oracle.sql [hunk 10]
```diff
@@ -1,3 +1,56 @@
...
+      		dbms_output.Put_line(command);
+     		EXECUTE IMMEDIATE command;
+    	END IF;
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/oracle.sql:23]: Fix indentation

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/oracle.sql [hunk 11]
```diff
@@ -1,3 +1,56 @@
...
+  		EXCEPTION
+    	WHEN NO_DATA_FOUND
+    	THEN
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/oracle.sql:26]: Fix indentation.

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/oracle.sql [hunk 12]
```diff
@@ -1,3 +1,56 @@
...
+  		SELECT a.constraint_name
+    		INTO con_name1
+	    	FROM all_cons_columns a
+	      		JOIN all_constraints c ON a.owner = c.owner AND a.constraint_name = c.constraint_name
+	    	WHERE
+	      	c.constraint_type = 'U' AND a.table_name = 'UM_USER' AND UPPER(a.OWNER) = UPPER(databasename)
+	      	AND a.column_name='UM_USER_NAME';
+
+	    IF TRIM(con_name1) IS NOT NULL THEN
+      		dbms_output.Put_line('Unique key (UM_USER_NAME,UM_TENANT_ID) is already exists');
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/oracle.sql:40]:
```suggestion
      		dbms_output.Put_line('Unique key (UM_USER_NAME,UM_TENANT_ID) already exists');
```

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/oracle.sql [hunk 13]
```diff
@@ -1,3 +1,56 @@
...
+
+    	EXCEPTION
+    	WHEN NO_DATA_FOUND
+    	THEN
+    	EXECUTE IMMEDIATE 'ALTER TABLE UM_USER ADD UNIQUE(UM_USER_NAME,UM_TENANT_ID)';
+  	END;
+ END;
+ /
+
+ALTER TABLE UM_USER ADD UNIQUE(UM_USER_ID)
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/oracle.sql:51]: Lets execute this after the relevant constraint check in the above block.

Comment (ashendes) [reply components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/oracle.sql:51]: Fix both instances.

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/postgresql.sql [hunk 14]
```diff
@@ -1,3 +1,28 @@
...
+    INTO con_name FROM information_schema.table_constraints AS tc
+    JOIN information_schema.key_column_usage AS kcu ON tc.constraint_name = kcu.constraint_name
+    JOIN information_schema.constraint_column_usage AS ccu ON ccu.constraint_name = tc.constraint_name
+    WHERE constraint_type ='UNIQUE' AND tc.table_name = 'um_user' AND kcu.column_name = 'um_user_id' GROUP BY tc.constraint_name ;
+
+    IF con_name IS NOT NULL THEN
+    	EXECUTE 'ALTER TABLE um_user DROP CONSTRAINT IF EXISTS '|| con_name || ';';
+    END IF;
+
+	ALTER TABLE um_user ADD CONSTRAINT um_user_um_user_id_key UNIQUE(um_user_id);
```
Comment (ashendes) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/postgresql.sql:15]: Fix indentation.

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql (hunk 1 above)
Comment (sandushi) [reply components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql]: resolved from https://github.com/wso2-extensions/identity-migration-resources/pull/243/commits/40c864ef2281d1ca8b38d0015a03cdc72763e33a

**Review by ashendes — APPROVED** [review]

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/db2.sql [hunk 15]
```diff
@@ -1,4 +1,34 @@
...
+BEGIN
+    DECLARE const_name VARCHAR(128);
+    DECLARE STMT VARCHAR(200);
+    SELECT  k.CONSTNAME
+    INTO const_name
+    FROM SYSCAT.KEYCOLUSE k JOIN SYSCAT.TABCONST t ON k.CONSTNAME = t.CONSTNAME AND k.TABSCHEMA = t.TABSCHEMA
+    WHERE t.TABNAME = 'UM_USER' AND t.TYPE='U' AND k.COLNAME = 'UM_USER_ID';
+    
+    IF const_name IS NOT NULL THEN
+        SET STMT = 'ALTER TABLE UM_USER DROP CONSTRAINT ' ||  const_name;
```
Comment (lashinijay) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/db2.sql:10]: shall we remove the extra space before `const_name`?

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/db2.sql [hunk 16]
```diff
@@ -1,4 +1,34 @@
...
+        PREPARE S1 FROM STMT;
+        EXECUTE S1;
+    END IF;
+    
+    SELECT  k.CONSTNAME
```
Comment (lashinijay) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/db2.sql:15]: Seems like there is an extra space after `SELECT` here also. Can you check other places also?

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql [hunk 17]
```diff
@@ -1,3 +1,33 @@
-CREATE INDEX IF NOT EXISTS INDEX_UM_USERNAME_UM_TENANT_ID ON UM_USER(UM_USER_NAME, UM_TENANT_ID);
+DROP ALIAS IF EXISTS ALTER_UM_USER;
+
+CREATE ALIAS ALTER_UM_USER AS $$ void alterUmUser(final Connection conn) throws SQLException {
+    String const_name = "";
+    PreparedStatement ps = conn.prepareStatement("SELECT tc.CONSTRAINT_NAME FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu JOIN INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc ON tc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME WHERE tc.TABLE_NAME ='UM_USER' AND tc.CONSTRAINT_TYPE = 'UNIQUE' AND kcu.COLUMN_NAME ='UM_USER_ID'");
+    ResultSet results =  ps.executeQuery();
+
+    while (results.next()) {
+        const_name  = results.getString("CONSTRAINT_NAME");
```
Comment (lashinijay) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql:9]: additional space before `=`

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql [hunk 18]
```diff
@@ -1,3 +1,33 @@
-CREATE INDEX IF NOT EXISTS INDEX_UM_USERNAME_UM_TENANT_ID ON UM_USER(UM_USER_NAME, UM_TENANT_ID);
+DROP ALIAS IF EXISTS ALTER_UM_USER;
+
+CREATE ALIAS ALTER_UM_USER AS $$ void alterUmUser(final Connection conn) throws SQLException {
+    String const_name = "";
+    PreparedStatement ps = conn.prepareStatement("SELECT tc.CONSTRAINT_NAME FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu JOIN INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc ON tc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME WHERE tc.TABLE_NAME ='UM_USER' AND tc.CONSTRAINT_TYPE = 'UNIQUE' AND kcu.COLUMN_NAME ='UM_USER_ID'");
+    ResultSet results =  ps.executeQuery();
```
Comment (lashinijay) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql:6]: additional space after `=`

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mysql.sql [hunk 19]
```diff
@@ -1,3 +1,27 @@
-CREATE INDEX INDEX_UM_USERNAME_UM_TENANT_ID ON UM_USER(UM_USER_NAME, UM_TENANT_ID);
+DROP PROCEDURE IF EXISTS ALTER_UM_USER;
+
+DELIMITER $$
+CREATE PROCEDURE ALTER_UM_USER()
+BEGIN
+    IF EXISTS(SELECT * FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA=DATABASE() AND TABLE_NAME='UM_USER') THEN
+    /* Drop the unique key on (UM_USER_ID,UM_TENANT_ID) with Constraint name UM_USER_ID */
```
Comment (bhagyasakalanka) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mysql.sql:7]: Indentation issue. Shall align the comment with the below if block.

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mysql_cluster.sql [hunk 20]
```diff
@@ -1,3 +1,27 @@
+DROP PROCEDURE IF EXISTS ALTER_UM_USER;
+
+DELIMITER $$
+CREATE PROCEDURE ALTER_UM_USER()
+BEGIN
+    IF EXISTS(SELECT * FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA=DATABASE() AND TABLE_NAME='UM_USER') THEN
+    /* Drop the unique key on (UM_USER_ID,UM_TENANT_ID) with Constraint name UM_USER_ID */
```
Comment (bhagyasakalanka) [inline components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mysql_cluster.sql:7]: Indentation issue. Shall align the comment with the below if block.

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/db2.sql (hunk 15 above)
Comment (sandushi) [reply components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/db2.sql:10]: Resolved by https://github.com/wso2-extensions/identity-migration-resources/pull/273/commits/ef1e71a38a1eaa983d51f30bf83bb38558e7d4ec

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/db2.sql (hunk 16 above)
Comment (sandushi) [reply components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/db2.sql:15]: Resolved by https://github.com/wso2-extensions/identity-migration-resources/pull/273/commits/ef1e71a38a1eaa983d51f30bf83bb38558e7d4ec

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql (hunk 17 above)
Comment (sandushi) [reply components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql:9]: Resolved by https://github.com/wso2-extensions/identity-migration-resources/pull/273/commits/ef1e71a38a1eaa983d51f30bf83bb38558e7d4ec

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql (hunk 18 above)
Comment (sandushi) [reply components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql:6]: Resolved by https://github.com/wso2-extensions/identity-migration-resources/pull/273/commits/ef1e71a38a1eaa983d51f30bf83bb38558e7d4ec

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql [hunk 21]
```diff
@@ -1,3 +1,33 @@
...
+    if (!const_name.equals("")) {
+        ps = conn.prepareStatement("ALTER TABLE UM_USER DROP CONSTRAINT " + const_name);
+        ps.execute();
+    }
+
+    ps = conn.prepareStatement("ALTER TABLE UM_USER ADD UNIQUE(UM_USER_ID)");
+    ps.execute();
+
+    ps = conn.prepareStatement("SELECT tc.CONSTRAINT_NAME FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu JOIN INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc ON tc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME WHERE tc.TABLE_NAME ='UM_USER' AND tc.CONSTRAINT_TYPE = 'UNIQUE' AND kcu.COLUMN_NAME ='UM_USER_NAME'");
+    results =  ps.executeQuery();
```
[the remark this answers is not in the document]
Comment (sandushi) [reply components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/h2.sql:20]: Resolved by https://github.com/wso2-extensions/identity-migration-resources/pull/273/commits/ef1e71a38a1eaa983d51f30bf83bb38558e7d4ec

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mysql.sql (hunk 19 above)
Comment (sandushi) [reply components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mysql.sql:7]: Resolved by https://github.com/wso2-extensions/identity-migration-resources/pull/273/commits/ef1e71a38a1eaa983d51f30bf83bb38558e7d4ec

File: components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mysql_cluster.sql (hunk 20 above)
Comment (sandushi) [reply components/org.wso2.is.migration/migration-resources/6.0.0/dbscripts/step5/um/mysql_cluster.sql:7]: Resolved by https://github.com/wso2-extensions/identity-migration-resources/pull/273/commits/ef1e71a38a1eaa983d51f30bf83bb38558e7d4ec

**Pull request merged**